crates/xline-client/src/clients/lock.rs

```diff
@@ -1,78 +1,247 @@
+            keep_alive: None,
+        }
+    }
+
+    /// Set session TTL.
+    #[inline]
+    #[must_use]
+    pub const fn with_ttl(mut self, ttl: i64) -> Self {
+        self.ttl = ttl;
+        self
+    }
+
+    /// Set session lease.
+    #[inline]
+    #[must_use]
+    pub const fn with_lease(mut self, lease_id: i64) -> Self {
+        self.lease_id = lease_id;
+        self
+    }
+
+    /// Crate an auto-renew session.
+    ///
+    /// # Errors
+    /// return errors when lease client grant failed
+    #[inline]
+    pub async fn build(mut self) -> Result<Self> {
+        let ttl = self.ttl;
+        self.lease_id = if self.lease_id == 0 {
+            let lease_response = self
+                .client
+                .lease_client
+                .grant(LeaseGrantRequest::new(ttl))
+                .await?;
+            lease_response.id
+        } else {
+            self.lease_id
+        };
+        let mut lease_client = self.client.lease_client.clone();
+        let lease_id = self.lease_id;
+        self.keep_alive = Some(tokio::spawn(async move {
+            /// The renew interval factor of which value equals 60% of one second.
+            const RENEW_INTERVAL_FACTOR: u64 = 600;
+            let (mut keeper, mut stream) = lease_client
+                .keep_alive(LeaseKeepAliveRequest::new(lease_id))
+                .await?;
+            loop {
+                keeper.keep_alive()?;
+                if let Some(resp) = stream.message().await? {
+                    if resp.ttl < 0 {
+                        return Err(XlineClientError::InvalidArgs(String::from(
+                            "lease keepalive response has negative ttl",
+                        )));
+                    }
+                    sleep(Duration::from_millis(
+                        resp.ttl.unsigned_abs().overflow_mul(RENEW_INTERVAL_FACTOR),
+                    ))
+                    .await;
+                }
+            }
+        }));
+        Ok(self)
     }
 }
 
-// These methods primarily originate from xline lock server,
-// see also: `xline/src/server/lock_server.rs`
-impl LockClient {
-    /// Creates a new `LockClient`
+impl Drop for Session {
     #[inline]
-    pub fn new(
-        curp_client: Arc<CurpClient>,
-        channel: Channel,
-        token: Option<String>,
-        id_gen: Arc<LeaseIdGenerator>,
-    ) -> Self {
+    fn drop(&mut self) {
+        if let Some(keep_alive) = self.keep_alive.take() {
+            keep_alive.abort();
+        }
+    }
+}
+
+/// Xutex（Xline Mutex） implements the sync lock with xline
+#[derive(Debug)]
+pub struct Xutex {
+    /// Lock session
+    session: Session,
+    /// Lock
+    prefix: String,
+    /// Lock key
+    key: String,
+    /// The revision of lock key
+    rev: i64,
+    /// Request header
+    header: Option<ResponseHeader>,
+}
+
+/// An RAII implementation of  a “scoped lock” of an `Xutex`
+#[derive(Default, Debug)]
+pub struct XutexGuard {
```

Comment:
The async_dropper requires the Default trait. If we introduce the lifetime bound for XutexGuard, the async_dropper may not work properly.

Reply:
How about we allow the user to do a manual unlock, and the unlock could return some information to the user.
For example:
```rust
{
    enum UnlockResult {
        Ok,
        // The lock may has been unlocked by the server
        Unsafe(Err),
    }
    let guard = xutex.lock_unsafe().await?;
    let result: UnlockResult = guard.unlock().await?;
    match result {
        // handles the lock result
        // ...
    }
}
```

Reply:
Manually unlocking may not be a good practice since the likelihood of a user forgetting to unlock is greater than the likelihood of a panic occurring.